### PR TITLE
lb api: moving load balancing policy specific configuration to extension configuration

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: lb api: load balancing policy specific configuration to extension configuration
Additional Description:
Risk Level: low. api only.
Testing: n/a.
Docs Changes: n/a.
Release Notes: n/a.
Platform Specific Features: n/a.

### CODE REVIEW

**Comment by wbpcode** [comment]
This is part of https://github.com/envoyproxy/envoy/issues/20634.

This PR moved the `zone_aware_config` and `locality_weighted_lb_config` from `common_lb_config`. to least request, round rabin and random's extension configuration.

And `consistent_hashing_lb_config` is moved to maglev's extension configuration.

**Comment by wbpcode** [comment]
cc @markdroth

is it possible for us to use `ConsistentHashingLbConfig` in the ring hash directly? Or to use `envoy.extensions.load_balancing_policies.common.v3.SlowStartConfig` in the configuration of least request or round ranbin directly?

I know we cann't update any API directly according to our API version policy even it's mared `not-implemented-hide`.
But seems that it would be better to use `envoy.extensions.load_balancing_policies.common.v3.SlowStartConfig` rather than `config.cluster.v3.Cluster.SlowStartConfig`.
And the ring hash should shared the same `ConsistentHashingLbConfig` with the maglev.

So, if it's possible to change the API directly (as a special case), it would be great.

File: api/envoy/extensions/load_balancing_policies/least_request/v3/least_request.proto
Comment (markdroth) [inline api/envoy/extensions/load_balancing_policies/least_request/v3/least_request.proto:59]: I guess this should also switch to `common.v3.SlowStartConfig`.  Technically, that's a breaking change, but I suspect no one is actually using this yet, so that's probably okay.

@temawi, please confirm that this change would not break anything for grpc-java.

File: api/envoy/extensions/load_balancing_policies/least_request/v3/least_request.proto
Comment (markdroth) [inline api/envoy/extensions/load_balancing_policies/least_request/v3/least_request.proto:25]: Why remove these extension comments?  I think they're needed to generate the right documentation.

File: api/envoy/extensions/load_balancing_policies/maglev/v3/maglev.proto
Comment (markdroth) [inline api/envoy/extensions/load_balancing_policies/maglev/v3/maglev.proto:31]: How about just inlining the two fields into this config?  That's what we did for the ring_hash config.  And that way, we don't need the `ConsistentHashingLbConfig` message at all.

File: api/envoy/extensions/load_balancing_policies/round_robin/v3/round_robin.proto
Comment (markdroth) [inline api/envoy/extensions/load_balancing_policies/round_robin/v3/round_robin.proto:25]: Same comment as above w.r.t. changing this to use `common.v3.SlowStartConfig`.

File: api/envoy/extensions/load_balancing_policies/least_request/v3/least_request.proto
Comment (markdroth) [inline api/envoy/extensions/load_balancing_policies/least_request/v3/least_request.proto:61]: Looks like we're repeating this `oneof` in 3 different places.  Would it make sense to define a new message in common.proto that contains this oneof, and then just have a single field of that new message type in each of these three places?

**Review by markdroth — COMMENTED** [review]
Thanks for doing this!

**Comment by markdroth** [comment]
Sorry, didn't see your questions above until after I'd reviewed.  Looks like we're thinking along the same lines, though.

> is it possible for us to use `ConsistentHashingLbConfig` in the ring hash directly? 

In my review, I had actually suggested the opposite, which is just inlining the fields in the maglev config, just like we did in the ring hash config.  That having been said, I am open to instead changing the ring_hash config to use `ConsistentHashingLbConfig` if we think it's likely that either (a) we will want to be able to add new fields to `ConsistentHashingLbConfig` in the future or (b) we think there will be more hash-based LB policies in the future that will want to share these fields.

In terms of API breakage policy, we can definitely add a new `ConsistentHashingLbConfig` field to the ring hash config if we want; the only question would be what to do about the existing inlined fields.  Technically, we can't remove them, but we can mark them as deprecated.  In practice, I suspect nothing would break if we simply removed them, since gRPC does not use those fields, but we don't really lose anything by leaving them there either.

> Or to use `envoy.extensions.load_balancing_policies.common.v3.SlowStartConfig` in the configuration of least request or round ranbin directly?

As per my review comments, this one is a bit trickier.  Technically, changing the type of the field is a breaking change.  However, I don't think gRPC uses these fields at all, and Envoy hasn't implemented these new extensions yet, so I don't think anything would break if we just made this change.

I'd be okay with allowing this as a one-off exception to our breakage policy, if none of the other @envoyproxy/api-shepherds objects.

File: api/envoy/extensions/load_balancing_policies/least_request/v3/least_request.proto
Comment (wbpcode) [reply api/envoy/extensions/load_balancing_policies/least_request/v3/least_request.proto:25]: These comments are wrong. The [#extension] comment should be complete name of extension.
I will ensure all these are correct when I implement these extensions.

File: api/envoy/extensions/load_balancing_policies/least_request/v3/least_request.proto
Comment (wbpcode) [reply api/envoy/extensions/load_balancing_policies/least_request/v3/least_request.proto:61]: Sounds great.

**Comment by wbpcode** [comment]
Thanks for you review comments. @markdroth 

> (a) we will want to be able to add new fields to ConsistentHashingLbConfig in the future or (b) we think there will be more hash-based LB policies in the future that will want to share these fields.

This is what I am worried about.

I will create new commit to address your comments.

File: api/envoy/extensions/load_balancing_policies/common/v3/common.proto
Comment (markdroth) [inline api/envoy/extensions/load_balancing_policies/common/v3/common.proto:25]: Would it make sense to nest `ZoneAwareLbConfig` and `LocalityWeightedLbConfig` inside of `LocalityLbConfig`?

**Review by markdroth — COMMENTED** [review]
This looks great to me!

File: api/envoy/extensions/load_balancing_policies/common/v3/common.proto
Comment (wbpcode) [reply api/envoy/extensions/load_balancing_policies/common/v3/common.proto:25]: > Looks like we're repeating this oneof in 3 different places. Would it make sense to define a new message in common.proto that contains this oneof, and then just have a single field of that new message type in each of these three places?

It's used to address previous comment. As said in the comment, the `oneof` repeat three times and this change eliminate this repeating.

Comment (markdroth) [reply api/envoy/extensions/load_balancing_policies/common/v3/common.proto:25]: Right, the `LocalityLbConfig` message is good.  What I'm suggesting here is to move the `ZoneAwareLbConfig` and `LocalityWeightedLbConfig` messages to be inside of `LocalityLbConfig`, since they're not used anywhere else.

Comment (wbpcode) [reply api/envoy/extensions/load_balancing_policies/common/v3/common.proto:25]: Hah, get it.

**Comment by wbpcode** [comment]
/assign-from @envoyproxy/senior-maintainers 

A LGTM is necessary for this API only PR.

**Comment by wbpcode** [comment]
/assign-from @envoyproxy/senior-maintainers

@htuch is out for vacation

**Comment by wbpcode** [comment]
cc @snowp could you give a LGTM to this API only PR when you have free time? Thanks.

File: api/envoy/extensions/load_balancing_policies/ring_hash/v3/ring_hash.proto
Comment (snowp) [inline api/envoy/extensions/load_balancing_policies/ring_hash/v3/ring_hash.proto:91]: Docs for this here and for the other lib configs?

**Review by snowp — CHANGES_REQUESTED** [review]
Makes sense to me, just one comment

**Review by snowp — APPROVED** [review]
Thanks!

**Pull request merged**
